src/IRPrinter.cpp

```diff
@@ -165,11 +170,89 @@ ostream &operator<<(ostream &stream, const Target &target) {
     return stream << "target(" << target.to_string() << ")";
 }
 
-namespace Internal {
+namespace {
+
+template<typename T>
+void emit_with_commas(ostream &stream, const std::vector<T> &v) {
+    stream << "(";
+    const char *sep = "";
+    for (const T &t : v) {
+        stream << sep << t;
+        sep = ", ";
+    }
+    stream << ")";
+};
+
+void print_func(ostream &stream, const Func &func, bool include_dependencies) {
+    using Internal::extract_rdom;
+    using Internal::ReductionDomain;
+    using Internal::simplify;
+
+    stream << "func " << func.name() << " = {\n";
+
+    // Topologically sort the functions
+    std::map<std::string, Internal::Function> env = include_dependencies
+        ? find_transitive_calls(func.function())
+        : std::map<std::string, Internal::Function>{{func.function().name(), func.function()}};
+    std::vector<std::string> order = include_dependencies
+        ? realization_order({ func.function() }, env).first
+        : std::vector<std::string>{func.function().name()};
+
+    for (int i = (int) order.size() - 1; i >= 0; i--) {
+        Func f(env[order[i]]);
+        for (int update_id = -1; update_id < f.num_update_definitions(); update_id++) {
+            std::vector<Expr> vals;
+            if (update_id >= 0) {
+                stream << "  " << f.name() << ".update[" << update_id << "]";
+                emit_with_commas(stream, f.update_args(update_id));
+                stream << " = ";
+                vals = f.update_values(update_id).as_vector();
+            } else {
+                stream << " " << f.name();
+                emit_with_commas(stream, f.args());
+                stream << " = ";
+                vals = f.values().as_vector();
+            }
+            if (vals.size() > 1) {
+                stream << "tuple<" << vals.size() << ">";
+            }
+            emit_with_commas(stream, vals);
+            // Assume that Tuples have the same (or no) RDom across all values.
+            ReductionDomain rdom = extract_rdom(vals.at(0));
```

Comment:
This assumption isn't true in general. An RVar may only be used in one val or even just in one arg (e.g. zeroing out a portion of a tensor). Also we really need the RDom name for this to be completely specified. Otherwise you can't distinguish RVars from Params.